tests/utils/test_uri.py

```diff
@@ -47,3 +48,146 @@ def test_uri_types():
     assert not is_http_uri("file://whatever")
     assert not is_http_uri("databricks://whatever")
     assert not is_http_uri("mlruns")
+
+
+def validate_append_to_uri_path_test_cases(cases):
+    for input_uri, input_path, expected_output_uri in cases:
+        assert append_to_uri_path(input_uri, input_path) == expected_output_uri
+        assert append_to_uri_path(input_uri, *posixpath.split(input_path)) == expected_output_uri
+
+
+def test_append_to_uri_path_joins_uri_paths_and_posixpaths_correctly():
+    validate_append_to_uri_path_test_cases([
+        ("", "path", "path"),
+        ("", "/path", "/path"),
+        ("path", "", "path/"),
+        ("path", "subpath", "path/subpath"),
+        ("path/", "subpath", "path/subpath"),
+        ("path/", "/subpath", "path/subpath"),
+        ("path", "/subpath", "path/subpath"),
+        ("/path", "/subpath", "/path/subpath"),
+        ("//path", "/subpath", "//path/subpath"),
+        ("///path", "/subpath", "///path/subpath"),
+        ("/path", "/subpath/subdir", "/path/subpath/subdir"),
+        ("file:path", "", "file:path/"),
```

Comment:
it's interesting that for 1-level deep paths we return a trailing "/" and for deeper paths we don't. is that intentional?

Reply:
and more importantly, are there any discrepancies in downstream behaviors this difference can cause?

Reply:
This appears to be caused by the behavior of `posixpath.join()`. I tried the following cases out in Python2 and Python 3:
```
>>> import posixpath
>>> posixpath.join("base", "")
'base/'
>>> posixpath.join("base/", "")
'base/'
>>> posixpath.join("/base", "",)
'/base/'
```

Given that we were previously leveraging `posixpath.join()` for the FileStore and SqlAlchemyStore and that multiple artifact stores rely on it for URI path component construction, this behavior seems consistent with pre-existing behavior and is (hopefully) safe.